lib/address_geocoder.rb

```diff
@@ -2,114 +2,118 @@
-    @street          = opt[:street]
-    @countries       = YAML.load_file('lib/address_geocoder/countries.yaml')['countries']['country']
-    @refined_address = {
-      :country     => nil,
-      :state       => nil,
-      :city        => nil,
-      :postal_code => nil,
-      :street      => nil
-    }
+    @api_key          = opt[:api_key]
+    @country          = opt[:country]
+    @state            = opt[:state]
+    @city             = opt[:city]
+    @postal_code      = opt[:postal_code]
+    @street           = opt[:street]
+    @countries        = YAML.load_file('lib/address_geocoder/countries.yaml')['countries']['country']
   end
 
-  def validate_address
-    return refine_address
+  def valid_address?
+    call_google if values_have_changed
+    return success?
   end
 
-  private
-  def refine_address
-    if situation_has_changed
-      # 1. Initialize refined address
-      @refined_address[:country] = match_country
-
-      # 2. Loop through the levels (once one works break the loop)
-      get_format_levels.each do |level_of_search|
-
-        # 2.1 Make call to Google
-        attempts = 0
-        begin
-          @google_response = HTTParty.get(get_final_url(level_of_search))
-        rescue
-          attempts += 1
-          if attempts <= 5
-            retry
-          else
-            raise
-          end
-        end
+  def suggested_addresses
+    call_google if values_have_changed
+    return false unless success?
+    return parse_response(@google_response['results'][0]['address_components'])
+  end
 
-        # 2.2 If the address succeeded:
-        if (@google_response['status'] == "OK") && (@google_response['results'][0]['address_components'].length > 1)
+  private
+  def call_google
+    # 1 initialize former address
+    @former_address = {
+      city: @city,
+      street: @street,
+      country: @country,
+      postal_code: @postal_code,
+      state: @state
+    }
 
-          # 2.2.1 Parse the address and break loop
-          parse_response(@google_response['results'][0]['address_components'])
-          break
+    # 2 Loop through the levels (once one works break the loop)
+    get_format_levels.each do |level_of_search|
+
+      # 2.1 Make call to Google
+      attempts = 0
+      begin
+        @google_response = HTTParty.get(get_final_url(level_of_search))
+      rescue
+        attempts += 1
+        if attempts <= 5
+          retry
+        else
+          raise
         end
       end
+
+      # 2.2 Break if the address succeeded
+      break if (@google_response['status'] == "OK") && (@google_response['results'][0]['address_components'].length > 1)
     end
-    return @refined_address
+  end
+
+  def success?
+    return (@google_response['status'] == "OK") && (@google_response['results'][0]['address_components'].length > 1)
   end
 
   def match_country
-    @countries.each do |country|
-      if @country.in? country.values
-        return country
+    if @country && @country[/[a-zA-Z]/]
```

Comment:
besides hard-code the valid regex, we could make a constant under our gem namespace
 something like . `AddressGeocoder::ValidCountryAlpha2Regex` 
`if @country.present? && @country.valid?`